contracts/entry-point/src/execute.rs

```diff
@@ -166,14 +166,119 @@ pub fn execute_user_swap(
+        Swap::SwapExactCoinIn(swap) => {
+            // Validate swap operations
+            validate_swap_operations(&swap.operations, &remaining_coin.denom, &min_coin.denom)?;
+
+            // Get swap adapter contract address from venue name
+            let user_swap_adapter_contract_address =
+                SWAP_VENUE_MAP.load(deps.storage, &swap.swap_venue_name)?;
+
+            // Create the user swap message args
+            let user_swap_msg_args: SwapExecuteMsg = swap.into();
+
+            // Create the user swap message
+            let user_swap_msg = WasmMsg::Execute {
+                contract_addr: user_swap_adapter_contract_address.to_string(),
+                msg: to_binary(&user_swap_msg_args)?,
+                funds: vec![remaining_coin],
+            };
+
+            response = response
+                .add_message(user_swap_msg)
+                .add_attribute("action", "dispatch_user_swap_exact_coin_in");
+        }
+        Swap::SwapExactCoinOut(swap) => {
+            // Validate swap operations
+            validate_swap_operations(&swap.operations, &remaining_coin.denom, &min_coin.denom)?;
+
+            // Get swap adapter contract address from venue name
+            let user_swap_adapter_contract_address =
+                SWAP_VENUE_MAP.load(deps.storage, &swap.swap_venue_name)?;
+
+            // Calculate the swap coin out by adding the min coin amount to the total affiliate fee amount
+            let swap_coin_out = Coin {
+                denom: min_coin.denom,
+                amount: min_coin.amount.checked_add(total_affiliate_fee_amount)?,
+            };
+
+            // Query the swap adapter to get the coin in needed to obtain the min coin plus affiliates
+            let user_swap_coin_in = query_swap_coin_in(
+                &deps,
+                &user_swap_adapter_contract_address,
+                &swap,
+                &swap_coin_out,
+            )?;
+
+            // Verify the user swap in denom is the same as the denom received from the message to the contract
+            if user_swap_coin_in.denom != remaining_coin.denom {
+                return Err(ContractError::UserSwapCoinInDenomMismatch);
+            }
+
+            // Calculate refund amount to send back to the user
+            let refund_amount = remaining_coin
+                .amount
+                .checked_sub(user_swap_coin_in.amount)?;
+
+            // If refund amount gt zero, then create the refund message and add it to the response
+            if refund_amount > Uint128::zero() {
+                // Get the refund address from the swap
+                let to_address = swap
+                    .refund_address
+                    .clone()
+                    .ok_or(ContractError::NoRefundAddress)?;
+
+                // Validate the refund address
+                deps.api.addr_validate(&to_address)?;
+
+                // Create the refund message
+                let refund_msg = BankMsg::Send {
```

Comment:
so I realize this is what we agreed on(?) but I do think it's awkward that the UX for swap exact outs is that you get some dust on the swap chain.

Reply:
Let's bring it up in standup